### PR TITLE
Generate inheritance diagrams for models and submodels in the documentation

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

Adds a Sphinx autodoc extension that appends inheritance diagrams for models and submodels classes in their docstrings, using a dropdown from [sphinx-design](https://sphinx-design.readthedocs.io/en/pydata-theme/dropdowns.html), and [sphinx.ext.inheritance_diagram](https://www.sphinx-doc.org/en/master/usage/extensions/inheritance.html) – which uses `graphviz` to generate a unidirectional graph in `.SVG` format that displays derived classes for a particular class.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all`
- [x] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
The built docs can be checked here: https://pybamm--3074.org.readthedocs.build/en/3074/

File: .readthedocs.yaml [hunk 1]
```diff
@@ -13,6 +13,9 @@ sphinx:
 
 # Set the version of Python and other tools you might need
 build:
+  # Install Graphviz to build SVG files
+  apt_packages:
+    - "graphviz"
```
Comment (Saransh-cpp) [inline .readthedocs.yaml:18]: Will the documentation build fail if this is not installed? If so, we might want to add this to "Install from source" documentation.

Comment (agriyakhetarpal) [reply .readthedocs.yaml:18]: The build will not fail, but on a previous build where it wasn't installed, it did not generate the graphs from the `.. digraph::` directive—showing this warning message:

```
"readthedocs" WARNING: dot command 'dot' cannot be run (needed for graphviz output), check the graphviz_dot setting
```

Comment (agriyakhetarpal) [reply .readthedocs.yaml:18]: We had a mention of `graphviz` in the [Developer installation instructions](https://docs.pybamm.org/en/latest/source/user_guide/installation/GNU-linux.html#developer-install), adding it to the source installation instructions too

File: docs/sphinxext/inheritance_diagram.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        # check if the class is a model or submodel
+        if "pybamm.models" in cls_name:
+            # check if the class derives from another class
+            if not len(getmro(obj)) > 2:
+                # do nothing if it is not a derived class
+                return
+
+            # Append the inheritance diagram to the docstring
+            lines.append("\n")
+            lines.append(".. dropdown:: View an inheritance diagram for this model")
```
Comment (tinosulzer) [inline docs/sphinxext/inheritance_diagram.py:25]:
```suggestion
            lines.append(".. dropdown:: View inheritance diagram for this model")
```

**Review by tinosulzer — APPROVED** [review]
thanks, looks good

**Pull request merged**